core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala

```diff
@@ -264,16 +240,106 @@ class SimpleAclAuthorizer extends Authorizer with Logging {
   }
 
+  /**
+    * Safely updates the resources acls by ensureing reads and writes respect the expected zookeeper version.
+    * Continues to retry until it succesfully updates.
+    *
+    * @param resource the resource to change acls for
+    * @param getNewAcls function to transform existing acls to new acls
+    * @return boolean indicating if a change was made
+    */
+  private def updateResourceAcls(resource: Resource)(getNewAcls: Set[Acl] => Set[Acl]): Boolean = {
+    val path = toResourcePath(resource)
+
+    var currentVersionedAcls =
+      if(aclCache.contains(resource))
+        getAclsFromCache(resource)
+      else
+      getAclsFromZk(resource)
+    var newVersionedAcls = currentVersionedAcls
+    var writeComplete = false
+    while (!writeComplete) {
+      val newAcls = getNewAcls(currentVersionedAcls.acls)
+      val data = Json.encode(Acl.toJsonCompatibleMap(newAcls))
+      try {
+        val (updateSucceeded, updateVersion) =
+          if(!newAcls.isEmpty) {
+           zkUtils.conditionalUpdatePersistentPathIfExists(path, data, currentVersionedAcls.version)
+          } else {
+            trace(s"Deleting path for $resource because it had no acls remaining")
+            (deletePath(path, currentVersionedAcls.version), 0)
+          }
+        if(!updateSucceeded) {
+          trace(s"Failed to update acls for $resource. Used version ${currentVersionedAcls.version}. Reading data and retrying update.")
+          currentVersionedAcls = getAclsFromZk(resource);
+        }
+        newVersionedAcls = VersionedAcls(newAcls, updateVersion)
+        writeComplete = updateSucceeded
+      } catch {
+        case e: ZkNoNodeException =>
+          try {
+            debug(s"Node for $resource does not exist, attempting to create it.")
+            zkUtils.createPersistentPath(path, data)
+            newVersionedAcls = VersionedAcls(newAcls, 0)
+            writeComplete = true
+          } catch {
+            case e: ZkNodeExistsException =>
+              debug(s"Failed to create node for $resource because it already exists. Reading data and retrying update.")
+              currentVersionedAcls = getAclsFromZk(resource);
+          }
+      }
+    }
+
+    if(newVersionedAcls.acls != currentVersionedAcls.acls) {
+      debug(s"Updated acls for $resource to ${newVersionedAcls.acls} with version ${newVersionedAcls.version}")
+      updateCache(resource, newVersionedAcls)
+      updateAclChangedFlag(resource)
+      true
+    } else {
+      debug(s"Updated acls for $resource, no change was made")
+      updateCache(resource, newVersionedAcls) // Even if no change, update the version
+      false
+    }
+  }
+
+  private def deletePath(path: String, expectedVersion: Int): Boolean = {
+    try {
+      zkUtils.zkConnection.getZookeeper.delete(path, expectedVersion) // Workaround until zkClient supports versioned deletes
```

Comment:
Hmm, if we use the raw ZK api here directly, we will need to handle the ConnectionLossException which zkclient hides from us. Perhaps we should just patch zkclient to expose versioned deleted and have zkclient release a new version.